src/auction/src/auction_processor.mo

```diff
@@ -0,0 +1,100 @@
+import Iter "mo:base/Iter";
+import List "mo:base/List";
+import Nat "mo:base/Nat";
+import Nat64 "mo:base/Nat64";
+import Prim "mo:prim";
+
+import { matchOrders } "mo:auction";
+
+import Assets "./assets";
+import Credits "./credits";
+import Orders "./orders";
+import Users "./users";
+import T "./types";
+
+module {
+
+  public func processAuction(
+    assets : Assets.Assets,
+    credits : Credits.Credits,
+    users : Users.Users,
+    assetId : T.AssetId,
+    sessionsCounter : Nat,
+    trustedAssetId : T.AssetId,
+    performanceCounter : Nat32 -> Nat64,
+  ) {
+    let startInstructions = performanceCounter(0);
+    let assetInfo = assets.getAsset(assetId);
+    let mapOrders = func(orders : List.List<(T.OrderId, T.Order)>) : Iter.Iter<(Float, Nat)> = orders
+    |> List.toIter<(T.OrderId, T.Order)>(_)
+    |> Iter.map<(T.OrderId, T.Order), (Float, Nat)>(_, func(_, order) = (order.price, order.volume));
+
+    let (_, _, dealVolume, price) = matchOrders(mapOrders(assetInfo.asks.queue), mapOrders(assetInfo.bids.queue));
+    if (dealVolume == 0) {
+      assets.pushToHistory(Prim.time(), sessionsCounter, assetId, 0, 0.0);
+      return;
+    };
+
+    // process fulfilled asks
+    var dealVolumeLeft = dealVolume;
+    label b while (dealVolumeLeft > 0) {
+      let ?(orderId, order) = assets.peekOrder(assetInfo, #ask) else Prim.trap("Can never happen: list shorter than before");
```

Comment:
This line exemplifies a problem that I see when reading this code. Can we write an auction processor at a lower level, which does not know anything about asset ids? The auction processor gets passed in an order book  and it does not know for which asset the order book is.

Here, with assets.peekOrder we start from assetInfo over and over again in each loop iteration. That is weird and hard to read. In my opinion there should not be such a function assets.peekOrder because peekOrder is a function needed in the auction processor and the auction processor is an algorithm that does not need to know about asset ids. peekOrder should be a function of the order book.

Can we make a class OrderBook?

Reply:
Added helper class `OrderBookService`